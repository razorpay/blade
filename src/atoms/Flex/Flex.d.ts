export { default } from './Flex.native';
export * from './Flex.native';
