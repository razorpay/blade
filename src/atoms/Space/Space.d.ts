export { default } from './Space.native';
export * from './Space.native';
