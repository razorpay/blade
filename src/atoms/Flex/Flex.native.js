import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';

const Flex = styled(
  ({
    alignContent,
    alignItems,
    alignSelf,
    display,
    flex,
    flexBasis,
    flexDirection,
    flexGrow,
    flexShrink,
    flexWrap,
    justifyContent,
    order,
    children,
    style,
    ...props
  }) => React.cloneElement(children, { ...props, style: [style, children.props.style] }),
)`
  ${(props) => (props.alignContent ? `align-content: ${props.alignContent};` : '')};
  ${(props) => (props.alignItems ? `align-items: ${props.alignItems};` : '')};
  ${(props) => (props.alignSelf ? `align-self: ${props.alignSelf};` : '')};
  ${(props) => (props.flex ? `flex: ${props.flex};` : '')};
  ${(props) => (props.flexBasis ? `flex-basis: ${props.flexBasis};` : '')};
  ${(props) => (props.flexDirection ? `flex-direction: ${props.flexDirection};` : '')};
  ${(props) => (props.flexGrow ? `flex-grow: ${props.flexGrow};` : '')};
  ${(props) => (props.flexShrink ? `flex-shrink: ${props.flexShrink};` : '')};
  ${(props) => (props.flexWrap ? `flex-wrap: ${props.flexWrap};` : '')};
  ${(props) => (props.justifyContent ? `justify-content: ${props.justifyContent};` : '')};
`;

Flex.propTypes = {
  alignContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'stretch',
  ]),
  alignItems: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  alignSelf: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  flex: PropTypes.number,
  flexBasis: PropTypes.string,
  flexDirection: PropTypes.oneOf(['column-reverse', 'column', 'row-reverse', 'row']),
  flexGrow: PropTypes.number,
  flexShrink: PropTypes.number,
  flexWrap: PropTypes.oneOf(['nowrap', 'wrap-reverse', 'wrap']),
  justifyContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'space-evenly',
  ]),
  children: PropTypes.element.isRequired,
};

Flex.defaultProps = {
  display: 'flex',
  flexDirection: 'column',
};

export default Flex;
