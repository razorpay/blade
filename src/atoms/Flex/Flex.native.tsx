import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';

const Flex = styled(
  ({
    alignContent,
    alignItems,
    alignSelf,
    display,
    flex,
    flexBasis,
    flexDirection,
    flexGrow,
    flexShrink,
    flexWrap,
    justifyContent,
    order,
    children,
    style,
    ...props
  }) =>
    React.cloneElement(children, {
      ...props,
      style: [style, children.props.style].filter(Boolean),
    }),
)`
  ${(props: any) => (props.alignContent ? `align-content: ${props.alignContent};` : '')};
  ${(props: any) => (props.alignItems ? `align-items: ${props.alignItems};` : '')};
  ${(props: any) => (props.alignSelf ? `align-self: ${props.alignSelf};` : '')};
  ${(props: any) => (props.flex ? `flex: ${props.flex};` : '')};
  ${(props: any) => (props.flexBasis ? `flex-basis: ${props.flexBasis};` : '')};
  ${(props: any) => (props.flexDirection ? `flex-direction: ${props.flexDirection};` : '')};
  ${(props: any) => (props.flexGrow ? `flex-grow: ${props.flexGrow};` : '')};
  ${(props: any) => (props.flexShrink ? `flex-shrink: ${props.flexShrink};` : '')};
  ${(props: any) => (props.flexWrap ? `flex-wrap: ${props.flexWrap};` : '')};
  ${(props: any) => (props.justifyContent ? `justify-content: ${props.justifyContent};` : '')};
`;

Flex.propTypes = {
  alignContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'stretch',
  ]),
  alignItems: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  alignSelf: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  flex: PropTypes.number,
  flexBasis: PropTypes.string,
  flexDirection: PropTypes.oneOf(['column-reverse', 'column', 'row-reverse', 'row']),
  flexGrow: PropTypes.number,
  flexShrink: PropTypes.number,
  flexWrap: PropTypes.oneOf(['nowrap', 'wrap-reverse', 'wrap']),
  justifyContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'space-evenly',
  ]),
  children: PropTypes.element.isRequired,
};

Flex.defaultProps = {
  display: 'flex',
  flexDirection: 'column',
};

export default Flex;
