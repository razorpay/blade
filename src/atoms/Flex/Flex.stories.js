import React from 'react';
import { storiesOf } from '@storybook/react';
import { select, text } from '@storybook/addon-knobs';
import Text from '../Text';
import View from '../View';
import Space from '../Space';
import Flex from './Flex';

const fillOptions = {
  row: 'row',
  column: 'column',
};

const justifyContentOptions = {
  'space-evenly': 'space-evenly',
  center: 'center',
  'flex-end': 'flex-end',
  'flex-start': 'flex-start',
  'space-around': 'space-around',
  'space-between': 'space-between',
};

storiesOf('Flex', module)
  .addParameters({
    component: Flex,
  })
  .add('default', () => (
    <Flex>
      <View>
        <Text>{text('Display Text 1', 'Flex child #1')}</Text>
        <Text>{text('Display Text 2', 'Flex child #2')}</Text>
      </View>
    </Flex>
  ))
  .add('flex justify-content', () => (
    <Flex
      justifyContent={select('justify-content', justifyContentOptions, 'center')}
      flexDirection="row"
    >
      <View>
        <Text>{text('Display Text 1', 'Flex child #1')}</Text>
        <Text>{text('Display Text 2', 'Flex child #2')}</Text>
      </View>
    </Flex>
  ))
  .add('Flex directions', () => (
    <Flex flexDirection={select('Direction', fillOptions, 'row')}>
      <View as="div">
        <Space margin={[1, 2]}>
          <View>Flex Child #1</View>
        </Space>
        <Space padding={[1, 2]}>
          <View>Flex Child #2</View>
        </Space>
        <Space padding={[1, 2]}>
          <View>Flex Child #3</View>
        </Space>
        <Space margin={[1, 2]}>
          <View>Flex Child #4</View>
        </Space>
      </View>
    </Flex>
  ))
  .add('Flex with spacing and children', () => (
    <Space margin={[1, 1]} padding={[1, 1]}>
      <Flex flexDirection="row">
        <View>
          <Space margin={[1, 1]} padding={[1, 1]}>
            <View>Flex Child #1</View>
          </Space>
          <Space margin={[1, 1]} padding={[1, 1]}>
            <View>Flex Child #2</View>
          </Space>
        </View>
      </Flex>
    </Space>
  ))
  .add('Flex flex', () => (
    <Flex flex={1}>
      <View as="div">
        <View>Flex Child #1</View>
        <View>Flex Child #2</View>
        <View>Flex Child #3</View>
        <View>Flex Child #4</View>
      </View>
    </Flex>
  ));
