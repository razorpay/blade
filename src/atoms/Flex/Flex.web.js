import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Flex = styled(
  ({
    alignContent,
    alignItems,
    alignSelf,
    display,
    flex,
    flexBasis,
    flexDirection,
    flexGrow,
    flexShrink,
    flexWrap,
    justifyContent,
    order,
    children,
    ...props
  }) => {
    if (React.Children.toArray(children).length !== 1) {
      throw new Error('Expected a single child for Space component');
    }
    return React.cloneElement(children, props);
  },
)`
  &&& {
    ${(props) => (props.alignContent ? `align-content: ${props.alignContent};` : '')};
    ${(props) => (props.alignItems ? `align-items: ${props.alignItems};` : '')};
    ${(props) => (props.alignSelf ? `align-self: ${props.alignSelf};` : '')};
    ${(props) => (props.flex ? `flex: ${props.flex};` : '')};
    ${(props) => (props.flexBasis ? `flex-basis: ${props.flexBasis};` : '')};
    ${(props) => (props.flexDirection ? `flex-direction: ${props.flexDirection};` : '')};
    ${(props) => (props.flexGrow ? `flex-grow: ${props.flexGrow};` : '')};
    ${(props) => (props.flexShrink ? `flex-shrink: ${props.flexShrink};` : '')};
    ${(props) => (props.flexWrap ? `flex-wrap: ${props.flexWrap};` : '')};
    ${(props) => (props.justifyContent ? `justify-content: ${props.justifyContent};` : '')};
    ${(props) => (props.display ? `display: ${props.display};` : '')};
  }
`;

Flex.propTypes = {
  alignContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'stretch',
  ]),
  alignItems: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  alignSelf: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  flex: PropTypes.number,
  flexBasis: PropTypes.string,
  flexDirection: PropTypes.oneOf(['column-reverse', 'column', 'row-reverse', 'row']),
  flexGrow: PropTypes.number,
  flexShrink: PropTypes.number,
  flexWrap: PropTypes.oneOf(['nowrap', 'wrap-reverse', 'wrap']),
  justifyContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'space-evenly',
  ]),
  children: PropTypes.element.isRequired,
};

Flex.defaultProps = {
  display: 'flex',
  flexDirection: 'row',
};

export default Flex;
