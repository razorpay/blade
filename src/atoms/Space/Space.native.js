import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';

import { makePxValue } from '../../_helpers/theme';

const Space = styled(({ margin, padding, children, ...props }) => {
  if (React.Children.toArray(children).length !== 1) {
    throw new Error('Expected a single child for Space component');
  }
  return React.cloneElement(children, props);
})`
  ${(props) => (props.margin ? `margin: ${makePxValue(props.margin)}` : '')};
  ${(props) => (props.padding ? `padding: ${makePxValue(props.padding)}` : '')};
`;

Space.propTypes = {
  margin: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string])),
  padding: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string])),
  children: PropTypes.element.isRequired,
};

export default Space;
