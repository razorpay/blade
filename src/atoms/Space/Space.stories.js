import React from 'react';
import styled from 'styled-components';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs';
import Text from '../Text';
import View from '../View';
import Space from './Space';

const StyledView = styled(View)`
  background-color: ${(props) => props.theme.colors.shade[940]};
`;

storiesOf('Space', module)
  .addParameters({
    component: Space,
  })
  .add('with margin and padding', () => (
    <Space margin={[1, 1]} padding={[2, 2]}>
      <StyledView>
        <Text
          //@ts-expect-error
          as="span"
        >
          {text('Display Text', 'The quick brown fox jumps over the lazy dog ')}
        </Text>
      </StyledView>
    </Space>
  ));
