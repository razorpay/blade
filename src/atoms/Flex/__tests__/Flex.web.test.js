import React from 'react';
import { render } from '@testing-library/react';

import { renderWithTheme } from '../../../_helpers/testing';
import View from '../../View';
import Text from '../../Text';
import Space from '../../Space';
import Flex from '../index';

describe('<Flex />', () => {
  describe('components composition', () => {
    it('renders child by applying flex and space properties', () => {
      const { container } = render(
        <Space>
          <Flex>
            <View>
              <View as="span">Flex Item #1</View>
              <View as="span">Flex Item #2</View>
            </View>
          </Flex>
        </Space>,
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe('flexDirection', () => {
    it('renders a child with flexDirection row', () => {
      const { container } = renderWithTheme(
        <Flex flexDirection="row">
          <View>
            <Space>
              <Text>Flex Item #1</Text>
            </Space>
            <Space>
              <Text>Flex Item #2</Text>
            </Space>
          </View>
        </Flex>,
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe('flex', () => {
    it('renders an inline styled child with flex property', () => {
      const viewStyles = { backgroundColor: 'red', height: 100 };
      const { container } = renderWithTheme(
        <Flex flex={1}>
          <Space>
            <View style={viewStyles}>
              <Text size="large">Flex Item #1</Text>
              <Text size="large">Flex Item #2</Text>
            </View>
          </Space>
        </Flex>,
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe('error', () => {
    beforeAll(() => jest.spyOn(console, 'error').mockImplementation());
    afterAll(() => jest.restoreAllMocks());
    it('should throw error when more than one nodes are passed as children', () => {
      const expectedErrorMessage = 'Expected a single child for Space component';
      expect(() =>
        render(
          <Flex>
            <View />
            <View />
          </Flex>,
        ),
      ).toThrow(expectedErrorMessage);
    });
  });
});
