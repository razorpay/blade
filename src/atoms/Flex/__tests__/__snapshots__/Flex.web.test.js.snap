// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flex /> components composition renders child by applying flex and space properties 1`] = `
<div>
  <div
    class="Viewweb__View-sc-1cphyjf-0 lhSUkp Flexweb__Flex-ni2cjg-0 dzXWge Spaceweb__Space-njatzl-0 jBqZrs"
  >
    <span
      class="Viewweb__View-sc-1cphyjf-0 lhSUkp"
    >
      Flex Item #1
    </span>
    <span
      class="Viewweb__View-sc-1cphyjf-0 lhSUkp"
    >
      Flex Item #2
    </span>
  </div>
</div>
`;

exports[`<Flex /> flex renders an inline styled child with flex property 1`] = `
<div>
  <div
    class="Viewweb__View-sc-1cphyjf-0 lhSUkp Spaceweb__Space-njatzl-0 jBqZrs Flexweb__Flex-ni2cjg-0 libPPx"
    style="background-color: red; height: 100px;"
  >
    <div
      class="Textweb__Text-sc-6jf9e4-0 jsGnSE"
      color="shade.980"
      data-testid="ds-text"
    >
      Flex Item #1
    </div>
    <div
      class="Textweb__Text-sc-6jf9e4-0 jsGnSE"
      color="shade.980"
      data-testid="ds-text"
    >
      Flex Item #2
    </div>
  </div>
</div>
`;

exports[`<Flex /> flexDirection renders a child with flexDirection row 1`] = `
<div>
  <div
    class="Viewweb__View-sc-1cphyjf-0 lhSUkp Flexweb__Flex-ni2cjg-0 dzXWge"
  >
    <div
      class="Textweb__Text-sc-6jf9e4-0 jsGnSE Spaceweb__Space-njatzl-0 jBqZrs"
      color="shade.980"
      data-testid="ds-text"
    >
      Flex Item #1
    </div>
    <div
      class="Textweb__Text-sc-6jf9e4-0 jsGnSE Spaceweb__Space-njatzl-0 jBqZrs"
      color="shade.980"
      data-testid="ds-text"
    >
      Flex Item #2
    </div>
  </div>
</div>
`;
